content/algorithms/Isomorphism/isomorphism.md

```diff
@@ -0,0 +1,221 @@
+
+Again we can detect that G and D are not isomorphic. But these conditions are not enough to say that 2 graphs are isomorphic. Let's look at the following example: 
+
+```{code-cell} ipython3
+plt.subplot(221)
+G = nx.Graph([(1, 2), (2, 3), (3, 4), (4, 1), (2,4)])
+nx.draw_circular(G, with_labels = True, node_color="g")
+plt.title("G", fontweight="bold")
+
+plt.subplot(222)
+H = nx.Graph([(1, 2), (2, 3), (3, 4), (2,4), (3, 1)])
+nx.draw_circular(H, with_labels = True, node_color = "c")
+plt.title("H", fontweight="bold")
+plt.show()
+```
+
+```{code-cell} ipython3
+nx.could_be_isomorphic(G1, G2)
+```
+
+These graphs meet all the necessary conditions but they're not isomorphic.
+
++++
+
+### Some classes of graphs with solution in polynomial time
+- Trees
+- Planar graphs(In fact, planar graph isomorphism is O(log(n)))
+- Interval graphs
+- Permutation graphs
+- Circulant graphs
+- Bounded-parameter graphs
+    - Graphs of bounded treewidth
+    - Graphs of bounded genus
+    - Graphs of bounded degree
+    - Graphs with bounded eigenvalue multiplicity
+    - k-Contractible graphs (a generalization of bounded degree and bounded genus)
+
++++
+
+# **Advanced Algorithms**
+
++++
+
+## **VF2**
+
+This algorithm is used to solve graph isomorphism and sub-graph isomorphism as well. 
+
+VF2 is a recursive algorithm where in each step we extend the current matching function to cover more nodes of both graphs until there are no more nodes to match. This is not a brute-force approach because there are some feasibility rules to avoid exploring the whole recursion tree. 
+
+
+Formally, We have a function $ M: s -> N(G) \times N(H) $. $M$ is a matching function between the subsets of nodes from $G$ and $H$ at the current state $s$. We start with an initial state $s_0$ with $M(s_0) = \emptyset$. In each step we consider a set of nodes to expand the current state $s$ to the following state $s'$. In this new state $M(s') = M(s) \cup {(g, h)} , g\in N(G), h\in N(H)$. The consistency condition is that the partial graphs $G$ and $H$ associated with $M(s)$ are isomorphic. There are two types of feasibility checks: 
+- syntactic (graph structure): consist of checking the consistency condition and also the k-look-ahead rules, for checking in advance if a consistent state $s$ has no consistent successors after k steps.
+- semantic(attributes). 
+
+Pseudocode:
+
++++
+
+
+Match(s)
+    
+    Input: Intermediate state
+    
+    Output: The mapping between the 2 graphs
+    
+    IF M(s) covers all nodes of H THEN:
+        RETURN M(s)
+    ELSE:
+        Compute P = {(g, h)...} the set of candidates for inclusion in M(s). 
+        FOR each p in P:
+            IF the feasibility rules succeed for the inclusion of p in M(s) THEN:
+                Compute the state of s'
+                MATCH(s')
+           ENDIF
+       ENDFOR
+       Restore data structures
+   ENDIF
+
++++
+
+**Time Complexity**
+
+
+- Best Case $\in \theta(n²)$
+- Wost Case $\in \theta(n!n)$
+
```

Comment:
Maybe we could add some information about the complexity? Like explain it a bit?